### ISSUE TITLE
Imported Solids aren't solid in a PostGIS-Backend

### ISSUE THREAD

**Issue by JaqucesT** [3dcitydb/importer-exporter#193]
I tried to calculate volumes of solids, generated by the current Importer/Exporter 4.3 and found, that PostGIS doesn't see those as solid.

Imported some official LoD1-Data from NRW Germany, that you can get here:
https://www.opengeodata.nrw.de/produkte/geobasis/3dg/lod1_gml/lod1_gml_paketiert/3d-gm_lod1_05120000_Remscheid_EPSG25832_CityGML.zip

CRSs name is:  urn.adv.crs.ETRS89_UTM32*DE_DHHN2016_NH*GCG2016

Setup the database with EPSG-Code 25832 and Height-EPSG-Code 7837
CRS name see above.

Imported all GML-Files of the Zip-File mentioned above successfully.

If i query the database about the validity of building-Surface-geometries, i get this result:

```sql
select
   b.id
  ,st_isvalid(a.geometry) as shouldBeValid
  ,b.objectclass_id
  ,c.classname
from
  citydb.surface_geometry a inner join
  citydb.cityobject b on a.cityobject_id = b.id inner join
  citydb.objectclass c on b.objectclass_id = c.id
order by
  b.id;
```

I find, that from the 6 surfaces that make up a simple LoD1-Solid, flat-roof and ground surface are valid objects, but all wall surfaces are invalid.

A solid generated by invalid surface-parts should be invalid too, and the query:

```sql
select
  st_issolid(b.geom)  as shouldBeSolid
from
  (select
    sg.solid_geometry as geom
  from
   citydb.surface_geometry sg
  where
    (not (sg.solid_geometry is null))
  ) b
where st_issolid(b.geom);
```

returns 0 records, - so all solids are "not Solid".

consequently ST_Volume doesn't want to calculate the volume of these objects.

PostGIS 3.1 and Postgresql 13 were used, database has all neccessary extensions incl. sfcgal.

Something wrong here?

Thanks for all kind of solutions and regards

Jaques

**Comment by BWibo** [comment]
Hey there, according to the [ST_VOLUME docs](https://postgis.net/docs/ST_Volume.html) you can call [ST_MakeSolid](https://postgis.net/docs/ST_MakeSolid.html) to cast the `solid_geometry` features as solids and enable volume calculation using SFCGAL.

**Comment by JaqucesT** [comment]
Hi BWibo,

tried covering the geom in st_makesolid() also, but that doesn't change the situation: all records still get evaluated as "not solid" (st_issolid == false).
Behaviour is sometimes a bit strange. Had files from another source, - still valid 3DCityGML-Files, where PostGIS allowed to calculate the volume, even with not solid solids, - result was always 0 then, - and after boxing those solids with st_makesolid, it wasn't possible to calculate volume any longer, - returned only error messages, reporting points, that didn't lie in the same plane. yes: st_volume possible with raw solids, giving always 0 as volume, after st_makesolid, st_volume throws errors.
From my point of view a bit strange.

Currently you also cannot calculate volume, when using an oracle backend. Imported those data into Oracle 12 / 18 / 19 with the same 4.3 Importer and while everything works fine in Oracle 12c, - 18c & 19c do not allow to calculate the volume either.
- Due to oracle they changed their internal handling of 3D of carthesian CRS. Documentation says, that you should be able to change this behaviour back to 12c-handling after patching your instance, but i wasn't able to get that running.

In PostGIS this error message: "Points don't lie in the same plane" and the invalidity of the wall-surfaces, are the culprints as i would think. Why are wall surfaces, that can get used after an import in oracle 12c, invalid in PostGIS? Why is it possible to calculate Volume with sdo_volume in oracle 12c from the same geometries, imported with the same Importer 4.3, but not it is not possible to do this in PostGIS?
It's the same set of coordinates, i don't think they get changed at the import, - resulting in valid solids in oracle but invalid in PostGIS.

**Comment by BWibo** [comment]
I did some tests and had problems with volume calculation as well. This allowed me to get some results:

```sql
SELECT sg.id, ST_VOLUME(ST_MakeSolid(sg.solid_geometry))
FROM surface_geometry sg, 
       (SELECT a.id
	FROM (SELECT sg.id, ST_isPlanar((ST_dump(sg.solid_geometry)).geom)::int planar
		   FROM surface_geometry sg) a
	GROUP BY a.id
	HAVING MIN(a.planar) = 1) b
WHERE b.id = sg.id
```
The query basically breaks down the `PolyhedralSurfaces` into `Polygons` which are checked for planarity using
`ST_isPlanar`. If there is at least one non-planar surface in the polyhedron, it is excluded from volume calculation.
The result is nt very satisfactory, as only a very small fraction of solids is fully free of polygons, that are recognized as non-planar in my dataset.

In my opinion this is a PostGIS/SFCGAL issue related to validity checks. Most GIS software allows to set tolerances for e.g. planarity checks. This is currently not supported by PostGIS/SFCGAL and I assume a very small tolerance value is used by default, that causes most solids/polygons to fail validity/planarity checks.

**Comment by clausnagel** [comment]
Thanks for testing, @BWibo. My assumption is also that the planarity check is too strict, and we would need to be able to pass a different tolerance value. Would triangulating the polygons in your query help?

**Comment by JaqucesT** [comment]
@BWibo
Were these LoD1 or LoD2 Data that you tested with?
I also found, that the error rate with LoD1 data is, after filtering out the non-planar datasets, quite low, but that changed dramatically with LoD2 data. With those datasets i found that only buildings with a flat roof had the same error rate as LoD1 data, - quite understandable from my point of view -, but i couldn't calculate the volume of buildings with other roof types.
The situation didn't change with a prior tesselation.

**Comment by BWibo** [comment]
I tested with LoD2 as well. The tesselation using [ST_Tesselate](https://postgis.net/docs/ST_Tesselate.html) is done by the SFCGAL backend as well. Hence, it is subject to the same validity checks and leads to the same planarity/validity errors as ST_Volume.

I did some more testing and found that ALL solids in my test dataset are no solids according to ST_isSolid and that they contain polygons with different orientation according to ST_Orientation. Fun fact, using the query from above involving the filtering with ST_isPlanar I am still able to compute volumes of 52 out of 316 solids in my test dataset.

```sql
SELECT a.id, a.count, b.issolid, c.vol
FROM 
	-- Count orientations per solid
	(SELECT sub.id, count(*) count
	FROM (SELECT DISTINCT sg.id, ST_Orientation((ST_DUMP(sg.solid_geometry)).geom)
		FROM surface_geometry sg
		WHERE sg.solid_geometry IS NOT NULL  
		GROUP BY sg.id) sub
	GROUP BY sub.id) a,
	-- ST_isSolid
	(SELECT sg.id, ST_isSolid(sg.solid_geometry) issolid
	FROM surface_geometry sg
	WHERE sg.solid_geometry IS NOT NULL
	ORDER BY issolid) b
LEFT JOIN 
  -- Filter solids with non-planar polygons and compute volume
  (SELECT sg.id, ST_VOLUME(ST_MakeSolid(sg.solid_geometry)) vol
	FROM surface_geometry sg, 
    (SELECT a.id
		FROM (
      SELECT sg.id, ST_isPlanar((ST_dump(sg.solid_geometry)).geom)::int planar
		  FROM surface_geometry sg) a
		GROUP BY a.id
		HAVING MIN(a.planar) = 1) b
	WHERE b.id = sg.id
	) c
ON b.id = c.id
WHERE a.id = b.id
```
According to OGC standards, orientation in Polyhedrons should be uniform RHR, right?
When I visualize the dataset with FME or FZKViewer, the oprientations seem fine (all normal point outwards).
Why is the volume calculation still working when running the planasity check? 
I can't really make sence of this...

**Comment by JaqucesT** [comment]
Hi BWibo,

i share your confusion.
Just a question:
In my case calculation of volumes where possible was extremly slow. To calculate the volume of 115 LoD1-Solids cost on a very current performant system 5 min time.
How was it with you?

**Comment by BWibo** [comment]
For me it worked much faster. But I tested on a very small dataset using a locally running [Docker container ](https://3dcitydb-docs.readthedocs.io/en/latest/3dcitydb/docker.html). You can use [`EXPLAIN ANALYZE`](https://www.postgresql.org/docs/13/sql-explain.html) to check e.g. if indices are used. It is possible that the bad performance is related to the query, not volume calculation.
If you used the queries I posted above, I just quickly put that togehter for testing. They are not performance optimized in any way.

**Comment by JaqucesT** [comment]
just found out, that it isn't st_volume that makes it all so slow for me: ST_3DArea to calculate the total surface area of the solid is the horribly slow function. From the 11 min time that this update takes on a 6 year old machine, 10 minutes are taken by st_3darea. st_3darea is not that slow, if you feed it with the solids right away, - but it doesn't produce anything usefull, only zero values. Calculations seem to be ok, when you first tesselate the solids, and that seems to take a huge amount of time.

**Comment by JaqucesT** [comment]
Does anybody know another way to calculate surface area & volume from the building geometries in citygml files? Can be a library for any programming language or anything else that let me do this for many files at once.
Searched from my point of view an eternety but couldn't find any open source solution, - well i know that fme does, but that's way to expensive. Thanks for any idea.

**Comment by clausnagel** [comment]
Ok, just did a simple test today as well. I set up a new 3DCityDB instance with 25832 as SRID. Afterwards, I imported the file `LoD1_370_5668_1_NW.gml` from the ZIP file referenced in the original post. I just tested with one file to keep my database small. After import, the table `surface_geometry` contains 1429 solid geometries in the `solid_geometry` column.

I first used the following SQL query to calculate volumes:

```sql
select st_volume(solid_geometry)
    from surface_geometry
    where solid_geometry is not null
```

This returns `0` for each solid geometry. So, I followed the example from the PostGIS documentation which uses `st_makesolid` in addition:

```sql
select st_volume(st_makesolid(solid_geometry))
    from surface_geometry
    where solid_geometry is not null
```

With this query, I get 1429 results and the query completes after 1 min 43 secs. All looks good to me. I'm running the 3DCityDB in a Docker container using the `latest` image from https://hub.docker.com/r/3dcitydb/3dcitydb-pg.

Screenshot from pgAdmin (just an excerpt):

![image](https://user-images.githubusercontent.com/5517439/125317486-55611980-e339-11eb-9f12-ddb5ff4c849f.png)

**Comment by clausnagel** [comment]
The following quote from the [PostGIS documentation](https://postgis.net/docs/ST_Volume.html) seems the explain the reason why we have to use `st_makesolid`:

> When closed surfaces are created with WKT, they are treated as areal rather than solid. To make them solid, you need to use ST_MakeSolid. Areal geometries have no volume.

Currently, the Importer/Exporter in fact imports the polyhedral surfaces as WKT because the PostGIS JDBC driver does not support 3D geometries natively. We have to investigate whether the way we import really makes using `st_makesolid` mandatory and if so, whether there is another way to do it. Until then, use `st_makesolid`.

**Comment by clausnagel** [comment]
One more comment on this: We could directly add a call to `st_makesolid` where we import the solid geometry:             https://github.com/3dcitydb/importer-exporter/blob/c8efa5f25109641550f5880008a2f9f335de2974/impexp-core/src/main/java/org/citydb/core/operation/importer/database/content/DBSurfaceGeometry.java#L139

According to the [PostGIS reference](https://postgis.net/docs/reference.html), `st_makesolid` just casts to a solid without performing checks. Thus, it should not affect import performance. The benefit would be that one can use `st_volume` (and other 3d functions) directly on the column `solid_geometry`. @yaozhihang, what do you think?

**Comment by JaqucesT** [comment]
@clausnagel 
Thank you, for your investigation.
There are two aspects:
First we have a database-column named "solid" that doesn't contain solids. From my point of view this is confusing, anyone would expect to find solids in those columns and not ToBe-Solids. 
Second the practical impact of solids, that cannot be used to do basic calculations like volume or surface area.
You tested with LoD1-Solids and as i wrote, with LoD1 solids, all having a flat roof, from my tests somewhere around 95% of imported GML-Datasets generate usable solids, - but even with 1 - 5 % invalid solids a regular update statement to calculate volume without filtering with st_isplanar() & st_isclosed() fails.
I found, and maybe i missed something here, that all solids LoD2-level, having other roof types as flat, cannot get used to calculate volume & surface area at all, - the mostly inclined roof-surfaces have all points not lying in the same plane and so you're stuck. This doesn't look as a problem of the importer/exporter but from sfcgal, as i would see it.

Wouldn't you expect automatically generated solids to be usable with all PostGIS-functions that support solids?
Maybe I'm an idealist but i would, - and frankly: what do solids help, that most functions declare invalid?
Wouldn't it be more straight forward to generate real,  fully usable solids where possible and keep that column empty (NULL) where the result would be an invalid geometry anyway?

**Comment by clausnagel** [comment]
Thanks for your opinions, @JaqucesT. I understand your points, but this is not a 3DCityDB issue in the first place.

- We store the solids with the correct PostGIS datatype in the database, and you can use them with any 3D PostGIS function when applying `ST_MakeSolid`.
- Automatically applying `ST_MakeSolid` when importing the data (as assumed in my previous post) would introduce a hard dependency on the SFCGAL backend. I am currently reluctant to introduce this dependency for the 3DCityDB because it's really simple to add `ST_MakeSolid` to your SQL statements. Something that needs more discussion.
- If the solids in the input data are invalid, we cannot fix them upon import. Crap in, crap out.
  - Still it can be useful to import invalid solids because then you can at least detect that they are invalid.
- If the PostGIS functions have issues (e.g. too strict tolerance values), these issues have to be reported to and fixed by the PostGIS team. For instance, I found a bug report for the [ST_IsSolid](https://trac.osgeo.org/postgis/ticket/4793) function which might also explain your results. I am not sure about the status of this ticket though.

**Comment by BWibo** [comment]
@clausnagel Thx for testing. I will have to do some more testing as well. It is interesting that the NRW data seems to work good. I'll have to try this with some original data from bavaria. My test dataset was possibly run through an FME workbench. As far as I can recall the workbench didn't touch the geometries, but maybe this causes the validity issues.
We are having the same problem with the dataset from Tokyo (avilable [here](https://wiki.tum.de/display/gisproject/Semantic+3D+City+Model+of+Tokyo)). This dataset required coordinate transformation from a geodetic system to a transformed system to fit in 3DCityDB. In that case the inaccuaracy of the transformation is enough to cause failed validity checks.

**Comment by BWibo** [comment]
According to this (https://gitlab.com/Oslandia/SFCGAL/-/releases/v1.4.1) there has been a fix in SFCGAL to `isPlane3D`.
This should be available in PostGIS 3.3 (see here https://postgis.net/docs/release_notes.html#idm44584)
Maybe this improves the situation regarding 3D operations like volume computation with failed validity tests.
I did not find the time to test this jet.

**Comment by thomashkolbe** [comment]
Well, while this issue is closed I would like to add one more thing I came across recently: the PostGIS function ST_IsValid() only looks at the 2D coordinates. Hence, any check of a vertical polygon (which looks like a line when the Z coordinate is ignored) or of a solid containing vertical polygons will fail. In fact, since 5 years there is a feature request for PostGIS to provide a ST_3DIsValid() function, but it seems that it has not been implemented so far: https://trac.osgeo.org/postgis/ticket/4363
